funsor/distributions.py

```diff
@@ -449,22 +450,81 @@ def eager_mvn(loc, scale_tril, value):
+        return None  # lazy
+    if not all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values()):
+        return None  # lazy
+
+    # Dovetail to avoid variable name collision in einsum.
+    equations1 = [''.join(c if c in ',->' else chr(ord(c) * 2 - ord('a')) for c in eqn)
+                  for _, eqn in coeffs.values()]
+    equations2 = [''.join(c if c in ',->' else chr(ord(c) * 2 - ord('a') + 1) for c in eqn)
+                  for _, eqn in coeffs.values()]
 
-    dim, = loc.output.shape
-    inputs, (loc, scale_tril) = align_tensors(loc, scale_tril)
-    inputs.update(value.inputs)
-    int_inputs = OrderedDict((k, v) for k, v in inputs.items() if v.dtype != 'real')
+    real_inputs = OrderedDict((k, v) for k, v in affine.inputs.items() if v.dtype == 'real')
+    assert tuple(real_inputs) == tuple(coeffs)
+
+    # Incorporate scale_tril before broadcasting.
+    eye = torch.eye(scale_tril.data.size(-1)).expand(scale_tril.data.shape)
+    prec_sqrt = Tensor(eye.triangular_solve(scale_tril.data, upper=False).solution,
+                       scale_tril.inputs)
+    tensors = []
+    for k, (coeff, eqn) in coeffs.items():
+        shape = (prec_sqrt.shape[-1], real_inputs[k].num_elements)
+        tensors.append((prec_sqrt @ coeff.reshape(shape)).reshape(coeff.shape))
+    tensors.append((prec_sqrt @ const.reshape((-1,))).reshape(const.shape))
+    tensors.append(scale_tril)
+
+    int_inputs, tensors = align_tensors(*tensors, expand=True)
+    coeffs, const, scale_tril = tensors[:-2], tensors[-2], tensors[-1]
+    batch_shape = const.shape[:-1]
+    dim = sum(d.num_elements for d in real_inputs.values())
 
-    precision = cholesky_inverse(scale_tril)
-    info_vec = precision.matmul(loc.unsqueeze(-1)).squeeze(-1)
+    info_vec = BlockVector(batch_shape + (dim,))
+    precision = BlockMatrix(batch_shape + (dim, dim))
+    offset1 = 0
+    for i1, (v1, c1) in enumerate(zip(real_inputs, coeffs)):
+        size1 = real_inputs[v1].num_elements
+        slice1 = slice(offset1, offset1 + size1)
+        inputs1, output1 = equations1[i1].split('->')
+        input11, input12 = inputs1.split(',')
+        info_vec[..., slice1] = torch.einsum(
+            f'...{input11},...{output1}->...{input12}', c1, const) \
+            .reshape(batch_shape + (size1,))
+        offset2 = 0
+        for i2, (v2, c2) in enumerate(zip(real_inputs, coeffs)):
+            size2 = real_inputs[v2].num_elements
+            slice2 = slice(offset2, offset2 + size2)
+            inputs2, output2 = equations2[i2].split('->')
+            input21, input22 = inputs2.split(',')
+            precision[..., slice1, slice2] = torch.einsum(
+                f'...{input11},...{input21}->...{input12}{input22}', c1, c2) \
+                .reshape(batch_shape + (size1, size2))
+            offset2 += size2
+        offset1 += size1
+
+    info_vec = info_vec.as_tensor()
+    precision = precision.as_tensor()
     log_prob = (-0.5 * dim * math.log(2 * math.pi)
```

Comment:
I think `dim` should be replaced by scale_tril.shape[-1].

Reply:
Done.